### PR TITLE
Add contribution guidelines

### PR DESCRIPTION
**Pull request by Nihlus**
This PR adds a set of contribution guidelines to complement the issue and pull request templates already added.

### CODE REVIEW

**Comment by Frassle** [comment]
Looks pretty good but is this not better suited to being on the github wiki vs being a file in the repo?

**Comment by Nihlus** [comment]
It's definitely good to have on the wiki as well, but if it's in the repo like this, a link to it is shown for new contributions asking them to review it before they submit stuff. It's a github integration that's quite nice, similar to the issue templates.

**Comment by Frassle** [comment]
Ok that sounds good then, I'll give it a proper read check for typos etc.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,268 @@
+# Contribution guidelines for OpenTK
+
+## Preface
+First of all, thank you for considering contributing to the OpenTK project! It's a large codebase
+with a lot of twists and turns, and a helping hand is always welcome.
+
+There are multiple ways to contribue to the project - creating bug reports, opening pull requests, 
+commenting on either and engaging in discussions about other contributions to name a few. This
```
Comment (Frassle) [inline CONTRIBUTING.md]: "either and" read oddly. Maybe just "commenting on and engaging in discussions"

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+document is intented as a set of guidelines to help your contribution get accepted faster, maintain
+a high standard, and to help us (the maintainers) set a few ground rules for working with us.
+
+If you have any questions about the contents of this document, the code, or how to contribute, please
+drop us a line on [Gitter](https://gitter.im/opentk/opentk). We'll be happy to answer as best we can.
```
Comment (Frassle) [inline CONTRIBUTING.md]: Discord as well?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+
+To make this at least somewhat easier, here's a few concrete general tips which you should stick to:
+
+* Always consider cross-platform gotchas.
+* Always work in small, iterative chunks which you can easily describe.
+* Avoid cosmetic or visual changes, unless your contribution is strictly focused on that.
+* Don't be afraid to ask, especially before diving in. There might be someone else working on the very
+same thing already!
+* Consider how your contribution might affect other contributions. Sometimes one change will break another
+if you're not careful.
```
Comment (Frassle) [inline CONTRIBUTING.md]: Maybe a point how it's acceptable (recommended even) to start pull requests with work still in progress. Just tag them as "WIP" to get early feedback.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+
+In terms of these guidelines, the terminology is as follows:
+* Must: If your contribution does not follow this rule, it will not be accepted.
+* Should: If your contribution does not follow this rule, it has a lower chance of being accepted.
+* May: If your contribution does not follow this rule, it's probably not going to matter that much. 
+It'd be a nice touch, though.
+
+With that in mind, check the following sections for more concrete and direct guidelines.
```
Comment (Frassle) [inline CONTRIBUTING.md]: Just "check the following sections for guidelines." the elaboration about concrete and direct isn't needed.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+$ git config --global user.signingkey QF3G6A39
+$ git config --global commit.gpgsign true
+```
+
+#### 4. Commit changes
+Once you've finished up a change, it's time to commit it. In doing so, you'll be writing some sort of 
+commit message, and there are some guidelines for how this should be formatted. Primarily, 
+
+* Keep the first line of the commit message 50 characters or less
+* If you need a longer description, keep the second line blank, and keep all subsequent more descriptive
```
Comment (Frassle) [inline CONTRIBUTING.md]: Second line should **always** be blank.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+
+* A clear, concise and descriptive title. As a rule of thumb, don't make it longer than twelve words or 
+72 characters.
+* A clear and detailed description of what the pull request has changed. This includes how the behaviour
+of the library will change if the pull request is accepted - a maintainer should be able to read your 
+description and fully understand what accepting it would mean without having to dive into the code.
+* Be based on the `develop` branch of the main OpenTK repository.
+
+All pull requests should have the following: 
+* If applicable, a compilable example which demonstrates the changes. A git repository is preferred, and
```
Comment (Frassle) [inline CONTRIBUTING.md]: hmm we don't do this. Maybe an example to opentk-examples instead?

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+your changed branch should be included as a submodule.
+
+A pull requests may have the following:
+* A short explanation of why you think these changes are neccesary, if it is not readily apparent from
```
Comment (Frassle) [inline CONTRIBUTING.md]: I think this is more of a should than a may.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+* If applicable, a compilable example which demonstrates the changes. A git repository is preferred, and
+your changed branch should be included as a submodule.
+
+A pull requests may have the following:
+* A short explanation of why you think these changes are neccesary, if it is not readily apparent from 
+the rest of the pull request.
+
+All code changes must follow these rules:
+
+* The [Style Guide](https://www.PLACE.HOLDER/style-guide) should be adhered to religiously. In general, this is
```
Comment (Frassle) [inline CONTRIBUTING.md]: PLACE HOLDER

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+Furthermore, if your new feature replaces or makes an existing feature obsolete, this must be clearly stated.
+This may prevent your pull request from being accepted in the current development cycle, or it may fast-track 
+it depending on the changes.
+
+### Cosmetic & Stylistic Changes
+Cosmetic and stylistic changes are those changes which do not affect executing code - that is, the library 
+operates exactly the same way before and after change, but the code might look nicer or follow the style guide 
+better.
+
+These types of pull requests are given less priority than others, and may have to wait some time.
```
Comment (Frassle) [inline CONTRIBUTING.md]: Probably true but I don't think it should be stated here.

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+* Must: If your contribution does not follow this rule, it will not be accepted.
+* Should: If your contribution does not follow this rule, it has a lower chance of being accepted.
+* May: If your contribution does not follow this rule, it's probably not going to matter that much. 
+It'd be a nice touch, though.
+
+With that in mind, check the following sections for more concrete and direct guidelines.
+
+## Setting Up
+For first-time contributors, there are a few steps that you'll need to go through in order
+to start contributing.
```
Comment (VPeruS) [inline CONTRIBUTING.md]: xunit and fsharp

File: CONTRIBUTING.md (hunk 6 above)
Comment (Nihlus) [reply CONTRIBUTING.md]: This isn't referring to examples that should stick around - just a "this is what changed, and this is how it works now"

File: CONTRIBUTING.md (hunk 8 above)
Comment (Nihlus) [reply CONTRIBUTING.md]: Yeah, we haven't written that one yet.

**Comment by Nihlus** [comment]
Updated.

File: CONTRIBUTING.md (hunk 10 above)
Comment (Nihlus) [reply CONTRIBUTING.md]: What do you mean?

Comment (VPeruS) [reply CONTRIBUTING.md]: point that it requires fsharp to run tests

**Comment by Nihlus** [comment]
@Frassle @VPeruS Any further comments or thoughts?

**Comment by Frassle** [comment]
I'll give it another read over tomorrow (getting a bit late now)

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,266 @@
+# Contribution guidelines for OpenTK
+
+## Preface
+First of all, thank you for considering contributing to the OpenTK project! It's a large codebase
+with a lot of twists and turns, and a helping hand is always welcome.
+
+There are multiple ways to contribue to the project - creating bug reports, opening pull requests,
```
Comment (VPeruS) [inline CONTRIBUTING.md]: contribue - typo?

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+#### 2. Create a working branch
+Development is done against the `develop` branch - this is where all the magic happens. Your
+changes should always be based on this branch, so in order to start working, create a new branch
+with an appropriate name and base it on `develop`.
+
+```bash
+$ git checkout -b my-branch -t origin/develop
+```
+
+#### 3. Let git know who you are
```
Comment (VPeruS) [inline CONTRIBUTING.md]: Maybe use this link: https://help.github.com/articles/set-up-git/ and move in on 1 position?

**Comment by VPeruS** [comment]
No entry about tests on fsharp and xunit

File: CONTRIBUTING.md (hunk 6 above)
Comment (Frassle) [reply CONTRIBUTING.md]: ok

File: CONTRIBUTING.md (hunk 8 above)
Comment (Frassle) [reply CONTRIBUTING.md]: Maybe just change it to say the files existing style should be followed and syntax rules are enforeced by StyleCop?

**Comment by Nihlus** [comment]
@VPeruS I don't think we need that - unit testing is mentioned, and we're trying to move away from F# tests.

**Comment by Frassle** [comment]
I think the only comments I have left are adding something about very early WIP pull requests and doing something about the Style Guide section.

**Comment by Nihlus** [comment]
Done.

**Pull request merged**